### PR TITLE
Handle external interface in `convert_call_args`

### LINKED ISSUE #1977: [SciPy]: asr_to_llvm: module failed verification. Error: Operand is null
```fortran
      subroutine dqc25f()
      external f
      call dqk15w(f)
      end
```

> LFortran

```console
% lfortran  /Users/hkalani/Desktop/scipy/scipy/integrate/quadpack/dqc25f.f --fixed-form --implicit-typing --implicit-interface --implicit-argument-casting --show-llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

define void @dqc25f() {
.entry:
  call void @dqk15w(<null operand!>)
  br label %return

return:                                           ; preds = %.entry
  ret void
}

declare void @dqk15w(float ()*)

declare float @f()
code generation error: asr_to_llvm: module failed verification. Error:
Operand is null
  call void @dqk15w(<null operand!>)
```

### LINKED ISSUE #2427: [SciPy]: AssertFailed for external functions passed as argument of interface
```fortran
subroutine a()
external f, g

call b(f, g)
end
```

> GFortran

```console
% gfortran a.f90       
Undefined symbols for architecture arm64:
  "_b_", referenced from:
      _a_ in ccPfzGu6.o
  "_f_", referenced from:
      _a_ in ccPfzGu6.o
  "_g_", referenced from:
      _a_ in ccPfzGu6.o
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
```

> GFortran -c
```
% gfortran a.f90 -c
```

> LFortran & LFortran -c --generate-object-code

```console
% lfortran a.f90 --implicit-interface --implicit-typing  --show-llvm
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 2128
    return emit_llvm(arg_file, lfortran_pass_manager,
  File "/Users/pranavchiku/repos/lfortran/src/bin/lfortran.cpp", line 843
    = fe.get_llvm(input, lm, pass_manager, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 291
    Result<std::unique_ptr<LLVMModule>> res = get_llvm2(code, lm, pass_manager, diagnostics);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 313
    diagnostics, lm.files.back().in_filename);
  File "/Users/pranavchiku/repos/lfortran/src/lfortran/fortran_evaluator.cpp", line 346
    compiler_options, run_fn, infile);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8966
    std::cout << LCompilers::pickle(asr, true, true, false) << std::endl;
  File "../libasr/asr.h", line 5037
  File "../libasr/asr.h", line 5013
  File "../libasr/asr.h", line 5038
  File "../libasr/asr.h", line 4747
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 880
    visit_symbol(*item.second);
  File "../libasr/asr.h", line 5040
  File "../libasr/asr.h", line 4756
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3551
    visit_procedures(x);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3746
    this->visit_stmt(*x.m_body[i]);
  File "../libasr/asr.h", line 5057
  File "../libasr/asr.h", line 4806
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 8140
    std::vector<llvm::Value *> args2 = convert_call_args(x, is_method);
  File "/Users/pranavchiku/repos/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 7621
    LCOMPILERS_ASSERT(llvm_symtab_fn_arg.find(h) != llvm_symtab_fn_arg.end());
AssertFailed: llvm_symtab_fn_arg.find(h) != llvm_symtab_fn_arg.end()
```

This happens becase `external f, g` declares `f`, `g`, in parent scope of `b` and hence somehow these are not present in `llvm_symtab_fn_arg.find`

### LINKED ISSUE #2748: calling external function as subroutine argument throws internal compiler error
```fortran
double precision function f(x)
  double precision x
  f = x**2
  return
end function f

subroutine sub(f)
  external f
  double precision f
  print *, f(2.0d0)
  return
end subroutine sub

recursive subroutine a()
external f 
call sub(f)
end

program main
  call a()
end
```
```console
$ gfortran b.f90 && ./a.out
   4.0000000000000000     
$ lfortran b.f90  --implicit-typing --implicit-interface --generate-object-code --use-loop-variable-after-loop  --legacy-array-sections
Internal Compiler Error: Unhandled exception
^C
```
Ref: https://github.com/lfortran/lfortran/issues/2714

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1977. Fixes #2748. Fixes #2427

### CODE REVIEW

File: tests/reference/llvm-external_03-c3442bb.stdout [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+declare void @_lfortran_printf(i8*, ...)
+
+define i32 @main(i32 %0, i8** %1) {
+.entry:
+  call void @_lpython_set_argv(i32 %0, i8** %1)
+  call void @b()
+  ret i32 0
+}
+
+declare void @_lpython_set_argv(i32, i8**)
```
Comment (Pranavchiku) [inline tests/reference/llvm-external_03-c3442bb.stdout:46]: Generated LLVM looks good to me.

**Comment by Pranavchiku** [comment]
This case do not work yet:
```fortran
function f(x) result (r)
   real r, x
   r = x
end function

subroutine `(f)
   interface
      function f(x) result (r)
         real r, x
      end function
   end interface
   real r
   r = f(2.0)
   print *, r
end subroutine

subroutine dqc25f()
   external f
   call dqk15w(f)
end

program main
   call dqc25f()
end program
```

Because, interface of `f` is updated in subroutine `dqk15w` , right now we only append to `changed_external_function_symbol` if it is external interface. It is easy fix.

File: tests/reference/llvm-external_03-c3442bb.stdout [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+define void @b() {
+.entry:
+  call void @a(float (float*)* @f.1)
+  br label %return
+
+return:                                           ; preds = %.entry
+  ret void
+}
+
+declare float @f.1(float*)
```
Comment (certik) [inline tests/reference/llvm-external_03-c3442bb.stdout:35]: I think this happens when it gets declared twice, correct?

File: tests/reference/llvm-external_03-c3442bb.stdout [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+return:                                           ; preds = %.entry
+  ret void
+}
+
+declare float @f(float*)
+
+define void @b() {
+.entry:
+  call void @a(float (float*)* @f.1)
```
Comment (certik) [inline tests/reference/llvm-external_03-c3442bb.stdout:28]: Here is it calling `f.1`, meaning that it won't be able to find the definition of this at link time, correct?

**Comment by certik** [comment]
Can you try to compile and link an example? I think it won't work, due to a missing `f.1` symbol.

**Comment by Pranavchiku** [comment]
This case works correctly: https://github.com/lfortran/lfortran/issues/2748#issue-1966068676

Here it has concrete definition of `f`

**Comment by certik** [comment]
How exactly does the `f.1` symbol get resolved? We need to understand that.

File: tests/reference/llvm-external_03-c3442bb.stdout (hunk 2 above)
Comment (Pranavchiku) [reply tests/reference/llvm-external_03-c3442bb.stdout:35]: Yes

**Comment by certik** [comment]
@Pranavchiku to move forward:

* can you add an integration test that fails in master but now works? Then we can merge this.
* As a separate investigation, we need to understand what is happening with the `f.1`.

**Comment by Pranavchiku** [comment]
I added the test, this PR can have a review.

**Review by certik — APPROVED** [review]
I think this looks good now, thanks!

**Pull request merged**